design/rest-api-documentation.md

```diff
@@ -0,0 +1,133 @@
+</dependency>
+```
+
+By spinning up the Keycloak server, developers should be able to send an HTTP request to http://localhost:8080/auth/openapi and get the OpenAPI document. Here is an example:
+
+```
+$ curl http://localhost:8080/auth/openapi
+---
+openapi: 3.0.1
+info:
+  title: Account REST API
+  contact:
+    name: Keycloak team
+    url: https://keycloak.org
+    email: keycloak-dev@googlegroups.com
+  license:
+    name: Apache 2.0
+    url: http://www.apache.org/licenses/LICENSE-2.0.html
+  version: v1alpha1
+servers:
+- url: /auth
+tags:
+- name: Account REST API
+  description: Account REST API operations.
+- name: OpenAPI Example
+  description: Get a text in various formats
+- name: Account REST service
+  description: REST API for the Account REST service
+paths:
+  /account/applications:
+    get:
+      tags:
+      - Account REST API
+      description: List of applications
+      parameters:
+      - name: name
+        in: query
+        schema:
+          type: string
+      responses:
+        "401":
+          description: Unauthorized, Failure
+        "500":
+          description: Error in getting the list of applications
+        "200":
+          description: Successful, List of applications
+```
+
+As we can see, with minimal effort it is possible to produce a JSON document that describes the functionalities of our API and can be used as a foundation for UI tools like Swagger to visualize and interact with our APIs. This document is served as a static file.
+
+## Swagger UI
+
+Swagger UI allows developers to visualize and generate documentation based on the OpenAPI specification. There are several ways of doing this:
+
+1. Inserting the URL in the [Swagger generator](https://generator.swagger.io/) which is available online. For example: http://localhost:8080/auth/openapi
+2. Spinning up a Docker image: 
+```
+docker pull swaggerapi/swagger-ui
+docker run -p 80:8080 swaggerapi/swagger-ui
+```   
+3. [Swagger Codegen](https://github.com/swagger-api/swagger-codegen#building). The best way to generate static HTML files for documentation.
+
+## Implementation plan
+
+1. Upgrade Keycloak distribution to WildFly 20.0.1.Final
+
+The current Keycloak distribution still use legacy feature packs from WildFly. In order to provide a subsystem xml file so that it can be used, we need the latest changes from [introduced in WildFly 20.0.1.Final](https://issues.redhat.com/browse/WFLY-13766).
+
+2. Introduce Microprofile OpenAPI subsystem in the Keycloak distributions
+3. Update the new Account REST API with the proper annotations
+4. Make sure that nothing is broken with the Quarkus distribution
+
+### Future work items
+
+The items below were included as nice to have because it's fairly easy to get these documents out of date and introduce it requires more discussion.
+
+* Generate static HTML files for the REST API
+* Generate static OpenAPI document and add to the release distribution
```

Comment:
Is there a Maven plugin available that can generate the OpenAPI document without starting the server?

Reply:
Not to my knowledge.  But I think there is an opportunity here to collaborate with the SmallRye OpenAPI project (Red Hat's implementation of the Microprofile-OpenAPI spec) to create one.  I believe it would be quite easy to integrate SmallRye OpenAPI into a maven plugin to generate the OpenAPI at build time rather than runtime.  Some of the Microprofile Specification features assume things are happening at runtime (e.g. filtering, model reader, etc) but the most important part (generating an OpenAPI document from annotated JAX-RS code) can be done at build time.  All we need in the SmallRye OpenAPI impl is a ShrinkWrap `Archive` with the compiled, annotated JAX-RS classes (and java beans).  So I think a maven plugin would be fairly trivial to write.

Reply:
Not that I know. Maybe what Eric suggested is something to consider in the future or discuss with Quarkus team.